go-sh: add sh package (includes shfmt and gosh)

mcmehrtens opened the pull request:
* adds a package for [sh](https://github.com/mvdan/sh), the software package housing `shfmt` and `gosh`
* package is named `go-sh` to avoid potential confusion with the Bourne shell
* `go-sh` creates two possible binaries: `shfmt` and `gosh`. I've dealt with this by adding variants for each.
  * To try and set sensible defaults, I've set the `shfmt` variant to default to true, since this seems to be the more widely used tool

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

mcmehrtens committed:
sh: add sh package (includes shfmt and gosh)

haampie commented:
I don't think this is a good name, given `/bin/sh`. Can you add some prefix, maybe `go-sh`? Or call it `shfmt`?

mcmehrtens committed:
go-sh: add prefix to the sh package

mcmehrtens commented:
> I don't think this is a good name, given `/bin/sh`. Can you add some prefix, maybe `go-sh`? Or call it `shfmt`?

@haampie, I addressed your comment in 2a356b28348ced3af7a75cd123fa92284cd0cccb.

I would vote on `go-sh`. If we went with `shfmt` as the package name and the user installed with `shfmt~shfmt+gosh`, it might seem a bit strange. It reads like, "`shfmt` is installed without `shfmt`."

Let me know if I should make any other changes :)

haampie commented on repos/spack_repo/builtin/packages/go_sh/package.py:
```diff
@@ -0,0 +1,57 @@
+    with when("+shfmt"):
+        sanity_check_is_file.append(join_path("bin", "shfmt"))
+    with when("+gosh"):
+        sanity_check_is_file.append(join_path("bin", "gosh"))
```
You cannot make class level code run "lazily" like that, the with body will always execute.

You'll have to do something like

```suggestion
    @property
    def sanity_check_is_file(self):
        files = []
        if self.spec.satisfies("+shfmt"):
            files.append(join_path("bin", "shfmt"))
        if self.spec.satisfies("+gosh"):
            files.append(join_path("bin", "gosh"))
        return files
```

mcmehrtens replied:
Shoot, I don't know how I missed that. Thank you! I applied your suggestion (1b60cd8f89c22e9899ec63744fce1718214dedf9).

haampie commented on repos/spack_repo/builtin/packages/go_sh/package.py:
```diff
@@ -0,0 +1,57 @@
+
+    variant("shfmt", default=True, description="Build and install shfmt")
+    variant("gosh", default=False, description="Build and install gosh")
+    conflicts("~shfmt~gosh", msg="One of shfmt or gosh must be specified")
```
why not both?

mcmehrtens committed:
Update repos/spack_repo/builtin/packages/go_sh/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>
